crypto/objects/obj_dat.c

```diff
@@ -294,9 +294,21 @@ static int ossl_obj_add_object(const ASN1_OBJECT *obj, int lock)
         if (ao[i] != NULL) {
             ao[i]->type = i;
             ao[i]->obj = o;
-            aop = lh_ADDED_OBJ_insert(added, ao[i]);
-            /* memory leak, but should not normally matter */
-            OPENSSL_free(aop);
+            aop[i] = lh_ADDED_OBJ_retrieve(added, ao[i]);
+            if (aop[i] != NULL)
+                aop[i]->type = -1;
+            (void)lh_ADDED_OBJ_insert(added, ao[i]);
```

Comment:
Can you elaborate on how this prevents a leak?  If the already-inserted gets its type set to -1, who is then responsible for freeing it when the last cached user is done with it?

Reply:
Everything gets freed here:
https://github.com/openssl/openssl/blob/186b3f6a016de8fcf8573be111e3d174ca20f1bc/crypto/objects/obj_dat.c#L215-L217
At this point `a->obj->nid` is used as a reference counter,
and the `a->obj->flags` is modified so that the object can be deleted,
before that point `ASN1_OBJECT_free(a->obj)` is a no-op.

Reply:
Ok, so on object add you're setting type to -1 so that future retrieve options won't match on it in the added_obj_cmp function.

does such a replacement happen often?  I ask because, if the number of type == -1 (e.g. stale or hidden) objects grows large, we would find ourselves in a position in which we were doing a large number of comparisons iterating through a list that would never match, leading to slowdowns.

I'm not opposed to this change, but it seems like it might be better to properly reference count elements in the hash table so that elements can be properly orphaned on removal while deferring the free operation until such time as the reference count reaches zero.  I understand that that is a big lift, but if we allow the caching of data that is owned by a hash table,  it seems like the right thing to do.

Reply:
No that replacement should only happen under very rare conditions.
One possible way how it could happen would be a race condition in `OBJ_create`.
When two threads call OBJ_create with different or NULL `oid` but identical `sn` or `ln`,
both could execute this check in parallel:
https://github.com/openssl/openssl/blob/758e9b537ac59680b4eaed77a81f5399cb38c0ae/crypto/objects/obj_dat.c#L778-L783
and won't see each other's conflicting names, as they are not yet installed.
And now both threads will execute the synchronized block here:
https://github.com/openssl/openssl/blob/758e9b537ac59680b4eaed77a81f5399cb38c0ae/crypto/objects/obj_dat.c#L795-L819
Furthermore, the first thread may even be able to get a pointer to the orphan object name
with `OBJ_nid2sn`, and store that for an arbitrary amount of time, so releasing that
orphan would be dangerous.